transport/graphsync/graphsync.go

```diff
@@ -149,19 +171,46 @@ func (t *Transport) OpenChannel(ctx context.Context,
 		exts = append(exts, doNotSendExt)
 	}
 
-	// Start tracking the data-transfer channel
 	ch := t.trackDTChannel(channelID)
 
+	// if we initiated this data transfer or the channel is open, run it immediately, otherwise
+	// put it in the request queue to rate limit DDOS attacks for push transfers
+	if channelID.Initiator == t.peerID || ch.isOpen {
```

Comment:
@hannahhoward Why do we need the `ch.isOpen` condition here  ?

@dirkmc What does `ch.Open` mean semantically for the data-transfer ?

Reply:
Once graphsync has called the outgoing requests hook, the channel is open (`isOpen` is true).

I think the idea here is that if the transfer is restarted, data-transfer will call `OpenChannel` again. So we don't want to put restarts at the back of the queue (`OpenChannel` will cancel the existing request and start a new one in its place)

Reply:
We may need to do a bit more refactoring here.
In the case where a transfer is restarted while the request is still in the queue we probably want to just replace the original request.

Reply:
@dirkmc 

You are right. So, if a queued transfer gets restarted, this PR will add a duplicate request to the queue, right ?

Reply:
Yes, for push requests I think it will add a duplicate request to the queue.